keps/sig-auth/3926-handling-undecryptable-resources/README.md

```diff
@@ -0,0 +1,952 @@
+(e.g. pods).
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+We need to make sure that a user that tries to perform an unconditional delete is well
+informed about the impact of what they are doing. This should be handled
+by one or more prompts from the `kubectl` client. In case of a plain HTTP DELETE
+request, at least issue a warning about what the impact of the action might be.
+Gate with an additional admission layer.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### Current state
+
+The encryption/decryption for encryption at rest is implemented via transformers
+that get applied to a resource in code that handles resource read/write from etcd3
+databases.
+
+The storage handling does not change with KMSv2, a resource transformer is provided
+in that case, too.
+
+References:
+1. [Encryption transformers creation](https://github.com/kubernetes/kubernetes/blob/8decaf3ae7f410ab3f3774f3895b9f3124b8a4c6/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L185)
+2. [Example of a resource retrieval with the transformer](https://github.com/kubernetes/kubernetes/blob/8decaf3ae7f410ab3f3774f3895b9f3124b8a4c6/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L155-L157)
+
+The code example 2. above shows that currently, when reading a resource fails, we
+lose all the context about the resource and a non-wrapping, generic internal error
+is returned.
+
+#### Retrieving a failing resource
+
+The current API errors don't appear to include an error status specific to storage. Therefore
+a new status should be introduced - `StatusReasonStoreReadError`.
+
+```go
+  // StatusReasonStoreReadError means that the server encountered an error while
+  // retrieving resources from the backend object store.
+  // This may be due to backend database error, or because processing of the read
+  // resource failed.
+  // Details:
+  //   "kind" string - the kind attribute of the resource being acted on.
+  //   "name" string - the prefix where the reading error(s) occurred
+  //   "causes" []StatusCause
+  //      - (optional):
+  //        - "type" CauseType - CauseTypeUnexpectedServerResponse
+  //        - "message" string - the error message from the store backend
+  //        - "field" string - the full path with the key of the resource that failed reading
```

Comment:
on a list there will be multiple keys, how will that be handled?

Reply:
"causes" are a slice, each key would get its own `StatusCause`

Reply:
How many will we include before deciding "message to large" and how will we represent that?

Reply:
That's a good question. I don't think that anything greater than 100 makes sense. I'll think about how to express the error truncation and add it to the design.